hnn_core/extracellular.py

```diff
@@ -0,0 +1,492 @@
+        self.imem_ptrvec = h.PtrVector(segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        self.imem_vec_len = int(self.imem_ptrvec.size())
+        self.imem_vec = h.Vector(self.imem_vec_len)
+
+        ptr_count = 0
+        for sec in secs_on_rank:
+            for seg in sec:  # section end points (0, 1) not included
+                # set Nth pointer to the net membrane current at this segment
+                self.imem_ptrvec.pset(
+                    ptr_count, sec(seg.x)._ref_i_membrane_)
+                ptr_count += 1
+        if ptr_count != self.imem_vec_len:
+            raise RuntimeError(f'Expected {self.imem_vec_len} imem pointers, '
+                               f'got {ptr_count}.')
+
+        # transfer resistances for each segment (keep in Neuron Matrix object)
+        self.r_transfer = h.Matrix(self.n_contacts, self.imem_vec_len)
+
+        for row, pos in enumerate(self.positions):
+            if self.method is not None:
+                transfer_resistance = list()
+                for sec, n_segs in zip(secs_on_rank, segment_counts):
+                    this_xfer_r = _transfer_resistance(sec, pos,
+                                                       sigma=self.sigma,
+                                                       method=self.method)
+                    transfer_resistance.extend(this_xfer_r)
+
+                self.r_transfer.setrow(row, h.Vector(transfer_resistance))
+            else:
+                # for testing, make a matrix of ones
+                self.r_transfer.setrow(row, h.Vector(segment_counts.sum(), 1.))
+
+        # record time for each array
+        self._ext_t = h.Vector().record(h._ref_t)
+
+        # contributions of all segments on this rank to total calculated
+        # potential at electrode (_PC.allreduce called in _simulate_dipole)
+        self._ext_v = h.Vector()
+        self.imem_vec = h.Vector(self.imem_vec_len)
+
+        # Attach a callback for calculating the potentials at each time step.
+        # Enables fast calculation of transmembrane current (nA) at each
+        # segment. Note that this will run on each rank, so it is safe to use
+        # the extra_scatter_gather-method, which docs say doesn't support
+        # "multiple threads".
+        cvode.extra_scatter_gather(0, self.calc_potential_callback)
+
+    @property
+    def n_samples(self):
+        """Return the length (in samples) of the extracellular data."""
+        return int(self._ext_v.size() / self.n_contacts)
+
+    @property
+    def voltages(self):
+        """The extracellular data (n_contacts x n_samples)."""
+        if len(self._ext_v) > 0:
+            # return as a Neuron Matrix object for efficiency
+            extmat = h.Matrix(self.n_contacts, self.n_samples)
+            extmat.from_vector(self._ext_v)
+            return extmat
+        else:
+            return None  # simulation not yet run
```

Comment:
raise error?

Reply:
Seemed excessively in-your-face. But too silent as it is?